### ISSUE TITLE
errors in P4_16 Example architecture description

### ISSUE THREAD

**Issue by tomrodeheffer** [p4lang/p4-spec#340]
In p4-16/spec/P4-16-spec.mdk section "Example architecture description"  there is the example:

```P4
parser Parser<IH>(packet_in b, out IH parsedHeaders);
// ingress match-action pipeline
control IPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         out OH outputHeaders,
                         out T toEgress,
                         out OutControl outCtrl);
// egress match-action pipeline
control EPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         in T fromIngress,
                         out OH outputHeaders,
                         out OutControl outCtrl);

control Deparser<OH>(in OH outputHeaders, packet_out b);
package Ingress<T, IH, OH>(Parser<IH> p,
                           IPipe<_, IH, OH> map,
                           Deparser<OH> d);
package Egress<T, IH, OH>(Parser<IH> p, Port
                          EPipe<_, IH, OH> map,
                          Deparser<OH> d);
package Switch<T>( // Top-level switch contains two packages
    // type types Ingress.IH and Egress.IH may be different
    Ingress<T, _, _> ingress,
    Egress<T, _, _> egress
);
```

This example contains several errors:

- On the `package Egress` line, the last word is `Port`.  This word should not be present.

- The definitions of `package Ingress` and `package Egress` are incorrect.
  - Notice that `package Ingress` does not use its type argument `T`.
  - Notice that `package Egress` does not use its type argument `T`.
  - Notice in `package Ingress` that the parameter `map` has type `IPipe<_, IH, OH>`.
  - Notice in `package Egress` that the parameter `map` has type `EPipe<_, IH, OH>`.
  - According to the description, `Ingress.IPipe` is supposed to be connected to `Egress.EPipe` via a channel that carries a value of type `T`.  As the example is written, the types `Ingress.IPipe.T` and `Egress.EPipe.T` are each arbitrary and therefore are not related.
  - The example should use `T` instead of `_` in these type parameters.

The description for this example refers to Figure `[#fig-switcharch]` which appears in the previous section.  The caption on this figure says "Switch architecture fragment implied by the previous set of declarations."  The caption is incorrect, because the declarations follow the figure, rather than being previous to it.  Actually, the figure would be positioned better following the example, instead of at its current rather unexpected location in the previous section.

**Comment by jnfoster** [comment]
Sigh. I just looked at the commit logs and this wasn't touched since the
initial translation from MS Word to Madoko.

Thanks for the bug report. We'll patch it up.

-N

**Comment by jnfoster** [comment]
See #349

**Comment by jafingerhut** [comment]
Looks like fixes for this issue may be committed now?

**Comment by jnfoster** [comment]
Indeed. Closed by #349.
